core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -298,8 +322,15 @@ public void runInContext() {
         // If the context has the effective deadline, we don't need to schedule an extra task.
         && !effectiveDeadline.equals(context.getDeadline())
         // If the channel has been terminated, we don't need to schedule an extra task.
-        && deadlineCancellationExecutor != null) {
-      deadlineCancellationFuture = startDeadlineTimer(effectiveDeadline);
+        && deadlineCancellationExecutor != null
+        // if already expired deadline let failing stream handle
+        && !(stream instanceof FailingClientStream)) {
+      deadlineCancellationNotifyApplicationFuture =
+          startDeadlineNotifyApplicationTimer(effectiveDeadline, observer);
+    } else {
+      if (cancellationListener instanceof ClientCallImpl.ContextCancellationListener) {
+        ((ContextCancellationListener) cancellationListener).setObserver(observer);
```

Comment:
You probably need to move line 320 (`context.addListener()`) to after this else-block, so that if the deadline is exceeded immediately, observer can still be notified.